### PR TITLE
Added support for nanosecond resolution Posix based wall clock time #311

### PR DESCRIPTION
**Pull request by peter-lawrey**

### CODE REVIEW

File: src/main/java/net/openhft/chronicle/core/time/PosixTimeProvider.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+package net.openhft.chronicle.core.time;
+
+import net.openhft.posix.ClockId;
+import net.openhft.posix.PosixAPI;
+
+/**
+ * This is the raw Posix time, if you need it to be unique or a multiple of 32 use a wrapper.
+ */
+public enum PosixTimeProvider implements TimeProvider {
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/core/time/PosixTimeProvider.java:27]: what is the difference between this and Chronicle-Ticker?

Comment (peter-lawrey) [reply src/main/java/net/openhft/chronicle/core/time/PosixTimeProvider.java:27]: Good question, let be review it.

Comment (RobAustin) [reply src/main/java/net/openhft/chronicle/core/time/PosixTimeProvider.java:27]: I've added @minborg for review as I think he suggests not using enums for utility methods like this.

**Review by RobAustin — APPROVED** [review]

File: src/main/java/net/openhft/chronicle/core/time/PosixTimeProvider.java (hunk 1 above)
Comment (minborg) [reply src/main/java/net/openhft/chronicle/core/time/PosixTimeProvider.java:27]: I think we should have a utility class that provides these different time providers and keep the actual implementations in an internal package. 

For example, `TimeProviders.acquirePosixTimeProvider()` or `TimeProviders.createUniqueMicroTimeProvider()`

Comment (minborg) [reply src/main/java/net/openhft/chronicle/core/time/PosixTimeProvider.java:27]: `SetTimeProvider` is using a non-enum pattern which is better from an API perspective. (But not as good as the above TImeProviders.xxxxx

image

Comment (minborg) [reply src/main/java/net/openhft/chronicle/core/time/PosixTimeProvider.java:27]: Are these classes to be used in conjunction with annotations? If not, we should really internalize them in the long run. We could move them to an internal package and deprecate the old ones (which could delegate form the internal ones and only act as a compatibility facade)

Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/core/time/PosixTimeProvider.java:27]: I thought we agreed at a dev meeting long ago we were going to continue to use enums for our singletons? We need to be consistent as right now we have different people doing different things

Comment (minborg) [reply src/main/java/net/openhft/chronicle/core/time/PosixTimeProvider.java:27]: Enums for singletons are fine with me. But they are best put in an internal package and exposed by other means.

**Pull request merged**